src/ivc/incrementally_verifiable_computation.rs

```diff
@@ -212,18 +215,108 @@ where
         })
     }
 
-    pub fn prove_step<const T: usize, RP1, RP2>(
+    pub fn fold_step<const T: usize, RP1, RP2>(
         &mut self,
-        _pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
-        _z0_primary: [C1::Scalar; A1],
-        _z0_secondary: [C2::Scalar; A2],
+        pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
     ) -> Result<(), Error>
     where
-        RP1: ROPair<C1::Scalar>,
-        RP2: ROPair<C2::Scalar>,
+        RP1: ROPair<C1::Scalar, Config = MainGateConfig<T>>,
+        RP2: ROPair<C2::Scalar, Config = MainGateConfig<T>>,
     {
-        // TODO #31
-        todo!("Logic at `RecursiveSNARK::prove_step`")
+        debug!("start fold step with folding 'secondary' by 'primary'");
+
+        debug!("start prove secondary trace");
+        let (secondary_new_trace, secondary_cross_term_commits) = nifs::vanilla::VanillaFS::prove(
+            pp.secondary.ck,
+            &self.secondary_nifs_pp,
+            &mut RP1::OffCircuit::new(pp.primary.params.ro_constant.clone()),
+            &self.secondary.relaxed_trace,
+            &self.secondary_trace,
+        )?;
+
+        debug!("prepare primary td");
+        // Prepare primary constraint system for folding
+        let (mut primary_td, primary_step_config) = Self::prepare_primary_td::<T, RP1>(
+            pp.primary.k_table_size,
+            [C1::Scalar::ZERO, C1::Scalar::ZERO], // TODO #154 #160
+        );
+        let primary_step_folding_circuit = StepFoldingCircuit::<'_, A1, C2, SC1, RP1::OnCircuit, T> {
+            step_circuit: &self.primary.step_circuit,
+            input: StepInputs::<'_, A1, C2, RP1::OnCircuit> {
+                step: C2::Base::from_u128(self.step as u128),
+                step_pp: &pp.primary.params,
+                public_params_hash: pp.digest().map_err(Error::WhileHash)?,
+                z_0: self.primary.z_0,
+                z_i: self.primary.z_i,
+                U: secondary_new_trace.U,
+                u: self.secondary_trace.u.clone(),
+                cross_term_commits: secondary_cross_term_commits,
+            },
+        };
+        debug!("start synthesize of 'step_folding_circuit' for primary");
+        self.primary.z_i = primary_step_folding_circuit.synthesize(
+            primary_step_config,
+            &mut SingleChipLayouter::<'_, C1::Scalar, _>::new(&mut primary_td, vec![])?,
+        )?;
+        debug!("start primary td postpone");
+        primary_td.postpone_assembly();
+
+        debug!("start generate primary plonk trace");
+        let primary_plonk_trace = VanillaFS::generate_plonk_trace(
+            pp.primary.ck,
+            &primary_td,
+            &self.primary_nifs_pp,
+            &mut RP2::OffCircuit::new(pp.secondary.params.ro_constant.clone()),
+        )?;
+
+        debug!("start prove primary trace");
+        let (primary_new_trace, primary_cross_term_commits) = nifs::vanilla::VanillaFS::prove(
```

Comment:
update it somewhere in this function:
`self.primary.relaxed_trace = primary_new_trace`